tests/phpunit/elementor/core/experiments/test-manager.php

```diff
@@ -566,6 +566,127 @@ public function test_validate_dependency__deactivates_an_experiment_when_its_dep
 		);
 	}
 
+	public function test_sort_allowed_options_by_dependencies() {
+		// Arrange.
+		$this->experiments->add_feature( [
+			'name' => 'test_A',
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_B',
+			'dependencies' => [
+				'test_A',
+				'non-existing-dep-that-should-be-ignored',
+			],
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_C',
+			'dependencies' => [ 'test_B' ],
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_D',
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_E',
+			'dependencies' => [ 'test_A' ],
+		] );
+
+		// Act.
+		$result = apply_filters(
+			'allowed_options',
+			[
+				'not-elementor' => [
+					'option_a',
+				],
+				'elementor' => [
+					'not_experiment',
+					'elementor_experiment-test_C',
+					'elementor_experiment-test_B',
+					'elementor_experiment-test_D',
+					'elementor_experiment-test_E',
+					'not_experiment_2',
+					'elementor_experiment-test_A',
+				],
+			]
+		);
+
+		// Assert.
+		$this->assertEqualSets(
+			[
+				'not-elementor' => [
+					'option_a',
+				],
+				'elementor' => [
+					'not_experiment',
+					'not_experiment_2',
+					'elementor_experiment-test_A',
+					'elementor_experiment-test_B',
+					'elementor_experiment-test_C',
+					'elementor_experiment-test_D',
+					'elementor_experiment-test_E',
+				],
+			],
+			$result
+		);
+
+		// Teardown.
+		$this->experiments->remove_feature( 'test_A' );
+		$this->experiments->remove_feature( 'test_B' );
+		$this->experiments->remove_feature( 'test_C' );
+		$this->experiments->remove_feature( 'test_D' );
+		$this->experiments->remove_feature( 'test_E' );
+	}
+
+	public function test_test_sort_allowed_options_by_dependencies__recursive_deps() {
+		// Arrange.
+		$this->experiments->add_feature( [
+			'name' => 'test_A',
+			'dependencies' => [ 'test_C' ],
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_B',
+			'dependencies' => [ 'test_A' ],
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_C',
+			'dependencies' => [ 'test_B' ],
+		] );
+
+		// Act.
+		$result = apply_filters(
+			'allowed_options',
+			[
+				'elementor' => [
+					'elementor_experiment-test_B',
+					'elementor_experiment-test_C',
+					'elementor_experiment-test_A',
+				],
+			]
+		);
+
+		// Assert.
+		$this->assertEqualSets(
+			[
+				'elementor' => [
+					'elementor_experiment-test_C',
+					'elementor_experiment-test_A',
+					'elementor_experiment-test_B',
+				],
+			],
+			$result
+		);
```

Comment:
hmm
do we need to fail for circular deps? what's the order logic in this case?

Reply:
?